src/shims/fs.rs

```diff
@@ -582,6 +597,256 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                } else {
+                    Ok(Scalar::Ptr(dir_ptr))
+                }
+            }
+            Err(e) => {
+                this.set_last_error_from_io_error(e)?;
+                Ok(Scalar::from_int(0, this.memory.pointer_size()))
+            }
+        }
+    }
+
+    fn readdir64_r(
+        &mut self,
+        dirp_op: OpTy<'tcx, Tag>,
+        entry_op: OpTy<'tcx, Tag>,
+        result_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("readdir64_r")?;
+
+        let dirp = this.force_ptr(this.read_scalar(dirp_op)?.not_undef()?)?;
+
+        let entry_ptr = this.force_ptr(this.read_scalar(entry_op)?.not_undef()?)?;
+        let dirent64_layout = this.libc_ty_layout("dirent64")?;
+        this.memory.check_ptr_access(
+            Scalar::Ptr(entry_ptr),
+            dirent64_layout.size,
+            dirent64_layout.align.abi,
+        )?;
+
+        if let Some(dir_iter) = this.machine.dir_handler.streams.get_mut(&dirp) {
+            match dir_iter.next() {
+                Some(Ok(dir_entry)) => {
+                    // write into entry, write pointer to result, return 0 on success
+                    let entry_place = this.deref_operand(entry_op)?;
+                    let ino64_t_layout = this.libc_ty_layout("ino64_t")?;
+                    let off64_t_layout = this.libc_ty_layout("off64_t")?;
+                    let c_ushort_layout = this.libc_ty_layout("c_ushort")?;
+                    let c_uchar_layout = this.libc_ty_layout("c_uchar")?;
+
+                    let name_offset = dirent64_layout.details.fields.offset(4);
+                    let name_ptr = entry_ptr.offset(name_offset, this)?;
+
+                    #[cfg(unix)]
+                    let ino = std::os::unix::fs::DirEntryExt::ino(&dir_entry);
+                    #[cfg(not(unix))]
+                    let ino = 0;
+
+                    #[cfg(unix)]
+                    let file_name = dir_entry.file_name();
+                    #[cfg(unix)]
+                    let file_name = std::os::unix::ffi::OsStrExt::as_bytes(file_name.as_os_str());
+                    #[cfg(not(unix))]
+                    let file_name = b"";
+
+                    let file_type = file_type_to_d_type(this, dir_entry.file_type())? as u128;
+
+                    let imms = [
+                        immty_from_uint_checked(ino, ino64_t_layout)?, // d_ino
+                        immty_from_uint_checked(0u128, off64_t_layout)?, // d_off
+                        immty_from_uint_checked(0u128, c_ushort_layout)?, // d_reclen
+                        immty_from_uint_checked(file_type, c_uchar_layout)?, // d_type
+                    ];
+                    this.write_packed_immediates(entry_place, &imms)?;
+                    this.memory.write_bytes(Scalar::Ptr(name_ptr), file_name.iter().copied())?;
```

Comment:
Ah so this is where `name_ptr` is relevant? (a) please group the name stuff together, currently this is pretty hard to decipher. Also some comments would be good to explain why the `write_packed_immediates` is not sufficient.

Reply:
Also, is `write_bytes` really right? Isn't this an OsString and should be written null-terminated via `write_os_str_to_c_str`?